File: python/Problem4.py
#!/usr/bin/env python
# -*- coding:utf-8 -*-
r"""
A palindromic number reads the same both ways. The largest palindrome made from the product of two 2-digit numbers is 9009 = 91  99.
Find the largest palindrome made from the product of two 3-digit numbers.
"""

import math


def is_palindrome(x):
    s = str(x)
    return s == s[::-1]


def is_3digits_product(x):
    for i in range(100, int(math.sqrt(x))):
        d, m = divmod(x, i)
        if not m and d < 1000:
            return True
    return False


if __name__ == '__main__':
    for i in reversed(xrange(10000, 998001)):
        if is_palindrome(i) and is_3digits_product(i):
            print(i)
            break

File: php/Problem4.php
<?php
/*
 * A palindromic number reads the same both ways. The largest palindrome made from the product of two 2-digit numbers is 9009 = 91  99.
 * Find the largest palindrome made from the product of two 3-digit numbers.
 */

function is_palindromic($x) {
    return $x == strrev($x);
}

function is_3digits_product($x) {
    for ($i = 100; $i <= sqrt($x); $i++) {
        $quot = $x / $i;
        $rem = $x % $i;
        if (!$rem && $quot < 1000) {
            return true;
        }
    }
    return false;
}

for ($i = 998001; $i >= 10000; $i--) {
    if (is_palindromic($i) && is_3digits_product($i)) {
        echo $i;
        break;
    }
}

